docs: MeshTrafficPermission (#1132)
kumahq/kuma-website: app/docs/dev/policies/meshtrafficpermission.md

---
title: MeshTrafficPermission (beta)
---

{% warning %}
This policy uses new policy matching algorithm and is in beta state,

> it should not be mixed with [TrafficPermission](../traffic-permissions.md).

Reviewer: This seems 
```suggestion
it should not be mixed with [TrafficPermission](../traffic-permissions).
```
I think now you shouldn't put the extension that's why checklink is broken

https://deploy-preview-1132--kuma.netlify.app/docs/dev/policies/meshtrafficpermission/
	404	https://deploy-preview-1132--kuma.netlify.app/docs/dev/policies/traffic-permissions.md

{% endwarning %}

## TargetRef support matrix

| TargetRef type    | top level | to  | from |
| ----------------- | --------- | --- | ---- |
| Mesh              | ✅        | ✅  | ❌   |
| MeshSubset        | ✅        | ✅  | ❌   |
| MeshService       | ✅        | ✅  | ❌   |

> | MeshServiceSubset | ✅        | ✅  | ❌   |

Reviewer: This seems wrong right? Isn't it supposed to be in the from column?

Author: Wow, good catch!

> | MeshGatewayRoute  | ✅        | ❌  | ❌   |

Reviewer 2: If there is no to/from then how is `MeshGatewayRoute` configured?

Reviewer 2: I think there is no way to configure it by [looking at the code](https://github.com/kumahq/kuma/blob/db9de2b90fabf25c4e146d77120fc3889cb98612/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1/meshtrafficpermission.go#L8-L17).

Author: Yeah, I think I had the same problem reading this matrix at first, like "if the top level is MeshGatewayRoute then to/from can't be set". While it should be read like "MeshGatewayRoute could be used only in the top level"

Author: I didn't understand the question initially, @slonka is right, today there is no way to use MeshTrafficPermission with MeshGatewayRoute. In the future, we'll be able to add this but only for cross-mesh. So I removed MeshGatewayRoute from matrix

---
title: MeshTrafficPermission (beta)
---

{% warning %}
This policy uses new policy matching algorithm and is in beta state,
it should not be mixed with [TrafficPermission](../traffic-permissions.md).
{% endwarning %}

## TargetRef support matrix

| TargetRef type    | top level | to  | from |
| ----------------- | --------- | --- | ---- |
| Mesh              | ✅        | ✅  | ❌   |
| MeshSubset        | ✅        | ✅  | ❌   |
| MeshService       | ✅        | ✅  | ❌   |
| MeshServiceSubset | ✅        | ✅  | ❌   |
| MeshGatewayRoute  | ✅        | ❌  | ❌   |

If you don't understand this table you should read [matching docs](/docs/{{ page.version }}/policies/matching).

## Configuration

### Action

> {{ site.title }} allows configuring one of 3 actions for a group of service's clients:

Reviewer: Should this be `site.mesh_product_name` (maybe just add it in `jekyll.yaml` like in #1122)

> * `ALLOW` - lets the requests through

Reviewer: ```suggestion
* `ALLOW` - allows incoming requests matching the from `targetRef`.
```

> * `DENY` - blocks all the requests

Reviewer: ```suggestion
* `DENY` - denies incoming requests matching the from `targetRef`
```

> * `ALLOW_WITH_SHADOW_DENY` - lets the requests through but emits logs as if requests were denied

Reviewer: ```suggestion
* `ALLOW_WITH_SHADOW_DENY` - same as `ALLOW` but will log as if request is denied, this is useful for rolling new restrictive policies without breaking things.
```

## Examples

### Service 'payments' allows requests from 'orders'

{% tabs allow-orders useUrlFragment=false %}
{% tab allow-orders Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  namespace: kuma-system
  name: allow-orders
  labels:
    kuma.io/mesh: default
spec:
  targetRef: # 1
    kind: MeshService
    name: payments
  from:
    - targetRef: # 2
        kind: MeshService
        name: orders
      default: # 3
        action: ALLOW
```

> Apply the configuration with `kubectl apply -f [..]`.

Reviewer: ```suggestion
```

{% endtab %}
{% tab allow-orders Universal %}

```yaml
type: MeshTrafficPermission
name: allow-orders
mesh: default
spec:
  targetRef: # 1
    kind: MeshService
    name: payments
  from:
    - targetRef: # 2
        kind: MeshService
        name: orders
      default: # 3
        action: ALLOW
```

> Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

Reviewer: ```suggestion
```

{% endtab %}
{% endtabs %}

#### Explanation

1. Top level `targetRef` selects data plane proxies that implement `payments` service.
    MeshTrafficPermission `allow-orders` will be configured on these proxies.

    ```yaml
    targetRef: # 1
      kind: MeshService
      name: payments
    ```

2. `TargetRef` inside the `from` array selects proxies that implement `order` service.
    These proxies will be subjected to the action from `default.action`.

    ```yaml
    - targetRef: # 2
        kind: MeshService
        name: orders

>     ```

Reviewer: Isn't there a whole thing with indent here?

Author: I'm not sure I follow what you mean, it's rendered as expected:
<img width="819" alt="image" src="https://user-images.githubusercontent.com/12110261/199802299-ab28fd89-6dc5-4c37-a583-13538d04a1e2.png">

Reviewer: Ok cool! I know there is a trick with nested blocks inside numbered list but if it renders well then we're all good!

3. The action is `ALLOW`. All requests from service `orders` will be allowed on service `payments`.

    ```yaml
    default: # 3
      action: ALLOW
    ```

### Deny all

{% tabs deny-all useUrlFragment=false %}
{% tab deny-all Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  namespace: kuma-system
  name: deny-all
  labels:

>     kuma.io/mesh: default

Reviewer: Let's remove the `default` mesh labels it's just overloading folks with more info than necessary

spec:
  targetRef: # 1
    kind: Mesh
  from:
    - targetRef: # 2
        kind: Mesh
      default: # 3
        action: DENY
```

> Apply the configuration with `kubectl apply -f [..]`.

Reviewer: ```suggestion
```

@michaelbeaumont suggested and I think it makes sense

{% endtab %}
{% tab deny-all Universal %}

```yaml
type: MeshTrafficPermission
name: deny-all
mesh: default
spec:
  targetRef: # 1
    kind: Mesh
  from:
    - targetRef: # 2
        kind: Mesh
      default: # 3
        action: DENY
```

> Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

Reviewer: ```suggestion
```

@michaelbeaumont suggested this and I think it makes sense

{% endtab %}
{% endtabs %}

#### Explanation

> 1. Top level `targetRef` selects all proxies in the mesh where is policy was applied (`default` mesh).

Reviewer: ```suggestion
1. Top level `targetRef` selects all proxies in the mesh.
```

{% tabs tags useUrlFragment=false %}
{% tab tags Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  namespace: kuma-system
  name: example-with-tags
  labels:
    kuma.io/mesh: default
spec:
  targetRef: # 1
    kind: Mesh
  from:
    - targetRef: # 2
        kind: MeshSubset
        tags:
          kuma.io/zone: us-east
      default: # 3
        action: ALLOW
    - targetRef: # 4
        kind: MeshSubset
        tags:
          env: dev
      default: # 5
        action: DENY
```

Apply the configuration with `kubectl apply -f [..]`.

{% endtab %}
{% tab tags Universal %}

```yaml
type: MeshTrafficPermission
name: example-with-tags
mesh: default
spec:
  targetRef: # 1
    kind: Mesh
  from:
    - targetRef: # 2
        kind: MeshSubset
        tags:
          kuma.io/zone: us-east
      default: # 3
        action: ALLOW
    - targetRef: # 4
        kind: MeshSubset
        tags:
          env: dev
      default: # 5
        action: DENY
```
Apply the configuration with `kumactl apply -f [..]` or with the [HTTP API](../../reference/http-api).

{% endtab %}
{% endtabs %}

#### Explanation

> 1. Top level `targetRef` selects all proxies in the mesh where is policy was applied (`default` mesh).

Reviewer: ```suggestion
1. Top level `targetRef` selects all proxies in the mesh.
```

    ```yaml
    targetRef: # 1
      kind: Mesh
    ```

2. `TargetRef` inside the `from` array selects proxies that have label `kuma.io/zone: us-east`.
    These proxies will be subjected to the action from `default.action`.

    ```yaml
    - targetRef: # 2
        kind: MeshSubset
        tags:
          kuma.io/zone: us-east
    ```

> 3. The action is `ALLOW`. All requests from the zone `us-east` will be allowed on all proxies in the `default` mesh.

Reviewer: ```suggestion
3. The action is `ALLOW`. All requests from the zone `us-east` will be allowed on all proxies.
```

    ```yaml
    default: # 3
      action: ALLOW
    ```

> 4. `TargetRef` inside the `from` array selects proxies that have label `kuma.io/zone: us-east`.

Reviewer: ```suggestion
4. `TargetRef` inside the `from` array selects proxies that have tags `kuma.io/zone: us-east`.
```

   These proxies will be subjected to the action from `default.action`.

    ```yaml
    - targetRef: # 4
        kind: MeshSubset
        tags:
          env: dev
    ```

> 5. The action is `DENY`. All requests from the env `dev` will be denied on all proxies in the `default` mesh.

Reviewer: ```suggestion
5. The action is `DENY`. All requests from the env `dev` will be denied on all proxies.
```

    ```yaml
    default: # 5
      action: DENY
    ```

{% tip %}
Order of rules inside the `from` array matters. 
Request from the proxy that has both `kuma.io/zone: east` and `env: dev` will be denied, 

> because the rule with `DENY` has bigger index in the `from` array. 

Reviewer: ```suggestion
This is because the rule with `DENY` is later in the `from` array than any `ALLOW` rules.
```
